docs/2019-12-16-Question-Mark-Variables-With-Ensures.md

```diff
@@ -0,0 +1,272 @@
+some of which may be deterministic, but there only exists a unique
+implementation of `randDifferentFromList(Is)`.
+Finally, note that `randDifferentFromList(Is)` is a _correct_ implementation
+for `fresh(Is)`.
+
+Here is the axiom of `fresh(Is)`
+
+```
+(FRESH) \forall Is:List{Int} . \exists I:Int . \neg( I \in Is) /\ (fresh(Is) -> * I)
+// equivalent form(s)
+//      \forall Is:List{Int} . fresh(Is) -> \exists I:Int . \neg( I \in Is) /\ * I
+//      \forall Is:List{Int} . fresh(Is) -> \exists I:Int . * (\neg( I \in Is) /\ I)
+//      \forall Is:List{Int} . fresh(Is) -> * \exists I:Int . \neg( I \in Is) /\ I
+```
+
+### Frontend Syntax
+
+We use the following K syntax to define `fresh(Is)`
+
+```
+syntax Exp ::= fresh (List{Int})
+rule fresh(Is:List{Int}) => ?I:Int
+  ensures notBool (?I inList{Int} Is)
+```
+
+## Example C: Arbitrary Number (Unspecific Function) `arb()`
+
+The function `arb()` is not a PL construct, but a mathematical function.
+Therefore, its axioms do not use rewrites but equalities.
+
+### Axioms
+
+The intended semantics of `arb()` is that it is an unspecified nullary function.
+The exact return value of `arb()` is unspecified in the semantics but up to the
+implementations.
+However, being a mathematical function, `arb()` must return the same value in
+any one implementation.
+
+The axiom of `arb()` is
+
+```
+(ARB) \exists I:Int . arb() = I
+//    this is exactly the (Function) axiom
+//    this is not equivalent to
+//      arb() = (\exists I:Int . I)
+```
+
+There are many variants of `arb()`. For example, `arbDifferentFromList(Is)` is
+a unspecified function whose return value must be different than those in `Is`.
+Here is its axiom
+
+```
+(ARBDIFF) \forall Is:List{Is} . \exists I:Int . \neg(I \in Is) /\ arbDifferentFromList(Is) = I
+```
+
+Note that `arbDifferentFromList(Is)` is different from `fresh(Is)`, because
+`fresh(Is)` _transitions_ to an integer not in `Is` (could be a different one
+each time it is used), while `arbDifferentFromList(Is)` _is equal to_ a (fixed)
+integer not in `Is`.
+
+### Skolemization
+
+One possible approach to implementing the abovewould be through [Skolemization](https://en.wikipedia.org/wiki/Skolem_normal_form),
+i.e., by replacing the existential variables with fresh uninterpreted functions
+applied to the universally quantified variables.
+
+For example, the `(ARBDIFF)` rule above could be Skolemized to
+
+```
+(ARBDIFF') \forall Is:List{Is} . \neg(arbDiffI(Is) \in Is) /\ arbDifferentFromList(Is) = arbDiffI(Is)
+```
+where
+```
+arbDiffI : List{Int} -> Int
+```
+
+is a symbol satisfying the `(Function)` axiom, without further constraints.
+
+### Frontend Syntax
+
+We do not need special frontend syntax to define `arb()`.
+We only need to define it in the usual way as a function
+(instead of a language construct), and provide no axioms for it
+(the `(Function)` axiom is generated automatically from the `function`
+attribute annotation.
+
```

Comment:
Yeah, I agree with that.  This is just saying that `arb(...)` is a function, so no need for anything special in terms of notation.